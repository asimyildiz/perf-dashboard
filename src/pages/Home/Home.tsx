import { FC, useEffect, memo } from 'react';
import { RouteComponentProps } from 'react-router-dom';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import { Helmet } from 'react-helmet';
import Navbar from 'react-bootstrap/Navbar';
import Nav from 'react-bootstrap/Nav';
import NavDropdown from 'react-bootstrap/NavDropdown';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';

import { AppState, AppThunk } from '../../store';
import { IVital } from '../../services/interfaces/vital.interface';
import { IDevice } from '../../services/interfaces/device.interface';
import { fromVital } from '../../interfaces/chartdata.interface';
import { fetchDeviceListIfNeed } from '../../store/device.list';
import { fetchVitalListIfNeed } from '../../store/vital.list';
import { groupBy, toMatrix } from '../../utils/helpers';
import { Chart } from '../../components';
import styles from './styles.module.scss';

export type Props = RouteComponentProps;

const Home: FC<Props> = (): JSX.Element => {
  const dispatch = useDispatch();
  const { deviceList, vitalList } = useSelector(
    (state: AppState) => state,
    shallowEqual,
  );

  // Fetch client-side data here
  useEffect(() => {
    dispatch(fetchDeviceListIfNeed());
  }, [dispatch]);

  const renderVitals = () => {
    const { readyStatus, vitals } = vitalList;

    if (!readyStatus || readyStatus === 'invalid' || readyStatus === 'request')
      return <p>Please select an url to display a web-site performance...</p>;

    if (readyStatus === 'failure') return <p>Oops, Failed to load list!</p>;

    const { devices } = deviceList;
    const keys: string[] = [];
    const grouped: Map<string, IVital[]> = groupBy(
      vitals,
      (vital: IVital) => vital.name,
    );
    grouped.forEach((_, key: string) => keys.push(key));

    const gridKeys: string[][] = toMatrix(keys, 2);
    return gridKeys.map((rowKeys: string[], index: number) => (
      <Row key={`row_${rowKeys[index]}`}>
        {rowKeys.map((colKey) => (
          <Col key={colKey} xs="6">
            <Chart
              title={colKey}
              devices={devices}
              items={fromVital(grouped.get(colKey))}
            />
          </Col>
        ))}
      </Row>
    ));
  };

  const onSelectUrl = (url: string | null): void => {
    const { readyStatus, devices } = deviceList;
    if (readyStatus === 'success' && devices.length > 0) {
      const ids: string[] = devices
        .filter((device: IDevice) => device.url === url)
        .map((device: IDevice) => device.id);
      dispatch(fetchVitalListIfNeed(ids));
    }
  };

  const renderUrls = () => {
    const { readyStatus, devices } = deviceList;

    if (
      !readyStatus ||
      readyStatus === 'invalid' ||
      readyStatus === 'request'
    ) {
      return <NavDropdown.Item href="#">Loading...</NavDropdown.Item>;
    }

    if (readyStatus === 'failure') {
      return (
        <NavDropdown.Item href="#">Could not fetch web-sites!</NavDropdown.Item>
      );
    }

    if (devices.length > 0) {
      const keys: string[] = [];
      const grouped = groupBy(devices, (device: IDevice) => device.url);
      grouped.forEach((_, key: string) => keys.push(key));

      return keys.map((key: string) => (
        <NavDropdown.Item
          key={`urls_${grouped.get(key)?.[0]?._id}`}
          eventKey={key}
        >
          {key}
        </NavDropdown.Item>
      ));
    }

    return (
      <NavDropdown.Item href="#">
        There are no web-sites reported!
      </NavDropdown.Item>
    );
  };

  return (
    <div className={styles.Home}>
      <Helmet title="Home" />
      <Navbar bg="light" expand="lg">
        <Navbar.Brand href="#home">Perf Dashboard</Navbar.Brand>
        <Navbar.Toggle aria-controls="basic-navbar-nav" />
        <Navbar.Collapse id="basic-navbar-nav">
          <Nav className="mr-auto" onSelect={onSelectUrl}>
            <NavDropdown title="Urls" id="basic-nav-dropdown">
              {renderUrls()}
            </NavDropdown>
          </Nav>
        </Navbar.Collapse>
      </Navbar>

      <Container fluid>{renderVitals()}</Container>
    </div>
  );
};

// Fetch server-side data here
export const loadData = (): AppThunk[] => [
  fetchDeviceListIfNeed(),
  // More pre-fetched actions...
];

export default memo(Home);
