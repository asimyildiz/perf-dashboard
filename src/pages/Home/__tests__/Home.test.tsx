import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

import { IConnection } from "../../../services/interfaces/connection.interface";
import { IDevice } from "../../../services/interfaces/device.interface";
import { IVital } from "../../../services/interfaces/vital.interface";
import { fetchDeviceListIfNeed } from "../../../store/device.list";
import mockStore from "../../../utils/mockStore";
import Home from "../Home";

describe("<Home />", () => {
  const renderHelper = (reducer = { readyStatus: "invalid" }) => {
    const { dispatch, ProviderWithStore } = mockStore({ deviceList: reducer, vitalList: reducer });
    const { container } = render(
      <ProviderWithStore>
        <MemoryRouter>
          {/* @ts-expect-error */}
          <Home />
        </MemoryRouter>
      </ProviderWithStore>
    );

    return { dispatch, firstChild: container.firstChild };
  };

  it("should fetch data when page loaded", () => {
    const { dispatch } = renderHelper();

    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(dispatch.mock.calls[0][0].toString()).toBe(
      fetchDeviceListIfNeed().toString()
    );
  });

  it("renders the loading status if data invalid", () => {
    expect(renderHelper().firstChild).toMatchSnapshot();
  });

  it("renders the loading status if requesting data", () => {
    const reducer = { readyStatus: "request" };

    expect(renderHelper(reducer).firstChild).toMatchSnapshot();
  });

  it("renders an error if loading failed", () => {
    const reducer = { readyStatus: "failure" };

    expect(renderHelper(reducer).firstChild).toMatchSnapshot();
  });

  it("renders the Chart Component if loading was successful", () => {
    const reducer = {
      readyStatus: "success",
      devices: [
        {
          _id: "60d0431a74a8240af878195e",
          id: "v1-1624261119060-2428637885473",
          url: "http://localhost/kasaba/",
          referrer: "",
          userAgent: "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.106 Safari/537.36",
          memory: 8,
          cpus: 8,
          connection: {
            effectiveType: "4g",
            rtt: 50,
            downlink: 10,
          } as IConnection,
          createdAt: "2021-06-21T07:43:22.700+00:00"
        } as IDevice,
      ],
      vitals: [
        {
          _id: "60d0431a74a8240af8781960",
          id: "v1-1624261119060-2428637885473",
          name: "FCP",
          delta: 1449.5999999940395,
          value: 1449.5999999940395,
        } as IVital,
        {
          _id: "60d0431a74a8240af8781961",
          id: "v1-1624261119060-2428637885473",
          name: "TTFB",
          delta: 353.7999999821186,
          value: 353.7999999821186,
        } as IVital,
        {
          _id: "60d0462a2fa99e0c816bcec7",
          id: "v1-1624261119060-2428637885473",
          name: "CLS",
          delta: 0.05607146357068885,
          value: 0.05607146357068885,
        } as IVital,
      ],
    };

    expect(renderHelper(reducer).firstChild).toMatchSnapshot();
  });

  it("not renders the Chart Component if loading was successful", () => {
    const reducer = {
      readyStatus: "success",
      devices: [],
      vitals: [
        {
          _id: "60d0431a74a8240af8781960",
          id: "v1-1624261119060-2428637885473",
          name: "FCP",
          delta: 1449.5999999940395,
          value: 1449.5999999940395,
        } as IVital,
        {
          _id: "60d0431a74a8240af8781961",
          id: "v1-1624261119060-2428637885473",
          name: "TTFB",
          delta: 353.7999999821186,
          value: 353.7999999821186,
        } as IVital,
        {
          _id: "60d0462a2fa99e0c816bcec7",
          id: "v1-1624261119060-2428637885473",
          name: "CLS",
          delta: 0.05607146357068885,
          value: 0.05607146357068885,
        } as IVital,
      ],
    };

    expect(renderHelper(reducer).firstChild).toMatchSnapshot();
  });
});
