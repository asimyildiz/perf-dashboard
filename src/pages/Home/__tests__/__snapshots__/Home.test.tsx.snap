// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> not renders the Chart Component if loading was successful 1`] = `
<div
  class="Home"
>
  <nav
    class="navbar navbar-expand-lg navbar-light bg-light"
  >
    <a
      class="navbar-brand"
      href="#home"
    >
      Perf Dashboard
    </a>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      class="navbar-toggler collapsed"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      class="navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        class="mr-auto navbar-nav"
      >
        <div
          class="nav-item dropdown"
        >
          <a
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle nav-link"
            href="#"
            id="basic-nav-dropdown"
            role="button"
          >
            Urls
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    class="container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col-6"
      >
        <div
          class="Chart"
        />
      </div>
      <div
        class="col-6"
      >
        <div
          class="Chart"
        />
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-6"
      >
        <div
          class="Chart"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Home /> renders an error if loading failed 1`] = `
<div
  class="Home"
>
  <nav
    class="navbar navbar-expand-lg navbar-light bg-light"
  >
    <a
      class="navbar-brand"
      href="#home"
    >
      Perf Dashboard
    </a>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      class="navbar-toggler collapsed"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      class="navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        class="mr-auto navbar-nav"
      >
        <div
          class="nav-item dropdown"
        >
          <a
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle nav-link"
            href="#"
            id="basic-nav-dropdown"
            role="button"
          >
            Urls
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    class="container-fluid"
  >
    <p>
      Oops, Failed to load list!
    </p>
  </div>
</div>
`;

exports[`<Home /> renders the Chart Component if loading was successful 1`] = `
<div
  class="Home"
>
  <nav
    class="navbar navbar-expand-lg navbar-light bg-light"
  >
    <a
      class="navbar-brand"
      href="#home"
    >
      Perf Dashboard
    </a>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      class="navbar-toggler collapsed"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      class="navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        class="mr-auto navbar-nav"
      >
        <div
          class="nav-item dropdown"
        >
          <a
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle nav-link"
            href="#"
            id="basic-nav-dropdown"
            role="button"
          >
            Urls
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    class="container-fluid"
  >
    <div
      class="row"
    >
      <div
        class="col-6"
      >
        <div
          class="Chart"
        />
      </div>
      <div
        class="col-6"
      >
        <div
          class="Chart"
        />
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-6"
      >
        <div
          class="Chart"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<Home /> renders the loading status if data invalid 1`] = `
<div
  class="Home"
>
  <nav
    class="navbar navbar-expand-lg navbar-light bg-light"
  >
    <a
      class="navbar-brand"
      href="#home"
    >
      Perf Dashboard
    </a>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      class="navbar-toggler collapsed"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      class="navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        class="mr-auto navbar-nav"
      >
        <div
          class="nav-item dropdown"
        >
          <a
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle nav-link"
            href="#"
            id="basic-nav-dropdown"
            role="button"
          >
            Urls
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    class="container-fluid"
  >
    <p>
      Please select an url to display a web-site performance...
    </p>
  </div>
</div>
`;

exports[`<Home /> renders the loading status if requesting data 1`] = `
<div
  class="Home"
>
  <nav
    class="navbar navbar-expand-lg navbar-light bg-light"
  >
    <a
      class="navbar-brand"
      href="#home"
    >
      Perf Dashboard
    </a>
    <button
      aria-controls="basic-navbar-nav"
      aria-label="Toggle navigation"
      class="navbar-toggler collapsed"
      type="button"
    >
      <span
        class="navbar-toggler-icon"
      />
    </button>
    <div
      class="navbar-collapse collapse"
      id="basic-navbar-nav"
    >
      <div
        class="mr-auto navbar-nav"
      >
        <div
          class="nav-item dropdown"
        >
          <a
            aria-expanded="false"
            aria-haspopup="true"
            class="dropdown-toggle nav-link"
            href="#"
            id="basic-nav-dropdown"
            role="button"
          >
            Urls
          </a>
        </div>
      </div>
    </div>
  </nav>
  <div
    class="container-fluid"
  >
    <p>
      Please select an url to display a web-site performance...
    </p>
  </div>
</div>
`;
